File: app/Models/Factura.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

class Factura extends Model
{
    use HasFactory;

    public const ESTADO_PENDIENTE = 'pendiente';
    public const ESTADO_BORRADOR = 'borrador'; // Legacy
    public const ESTADO_EMITIDA = 'emitida';
    public const ESTADO_ANULADA = 'anulada';

    public const ESTADOS = [
        self::ESTADO_PENDIENTE,
        self::ESTADO_EMITIDA,
        self::ESTADO_ANULADA,
    ];

    public const ESTADO_ALIASES = [
        self::ESTADO_BORRADOR => self::ESTADO_PENDIENTE,
    ];

    protected $fillable = [
        'codigo',
        'numero',
        'fecha',
        'ciudad_expedicion',
        'cliente_id',
        'cliente_nombre',
        'cliente_nit',
        'cliente_contacto',
        'cliente_direccion',
        'cliente_ciudad',
        'observaciones',
        'firma_path',
        'firma_nombre',
        'firma_cargo',
        'firma_empresa',
        'subtotal',
        'iva_total',
        'impuestos',
        'total',
        'estado',
        'created_by',
        'updated_by',
        'emitida_at',
        'emitida_by',
        'anulada_at',
        'anulada_by',
        'user_id',
    ];

    protected function casts(): array
    {
        return [
            'fecha' => 'date',
            'subtotal' => 'decimal:2',
            'iva_total' => 'decimal:2',
            'impuestos' => 'decimal:2',
            'total' => 'decimal:2',
            'emitida_at' => 'datetime',
            'anulada_at' => 'datetime',
        ];
    }

    public function scopeBuscar(Builder $query, string $buscar): Builder
    {
        $buscar = trim($buscar);

        if ($buscar === '') {
            return $query;
        }

        return $query->where(function (Builder $subQuery) use ($buscar) {
            $subQuery->where('codigo', 'like', "%{$buscar}%")
                ->orWhere('numero', 'like', "%{$buscar}%")
                ->orWhere('cliente_nombre', 'like', "%{$buscar}%")
                ->orWhere('cliente_nit', 'like', "%{$buscar}%");
        });
    }

    public function scopeEstado(Builder $query, ?string $estado): Builder
    {
        $estado = self::normalizeEstado($estado);

        if ($estado === null) {
            return $query;
        }

        return $query->where('estado', $estado);
    }

    public function scopeVisibleFlow(Builder $query): Builder
    {
        return $query->whereIn('estado', [
            self::ESTADO_PENDIENTE,
            self::ESTADO_BORRADOR, // Legacy
            self::ESTADO_EMITIDA,
            self::ESTADO_ANULADA,
        ]);
    }

    public function scopePendiente(Builder $query): Builder
    {
        return $query->whereIn('estado', [
            self::ESTADO_PENDIENTE,
            self::ESTADO_BORRADOR, // Legacy
        ]);
    }

    public static function normalizeEstado(?string $estado): ?string
    {
        if ($estado === null) {
            return null;
        }

        $normalized = mb_strtolower(trim($estado));

        if (isset(self::ESTADO_ALIASES[$normalized])) {
            return self::ESTADO_ALIASES[$normalized];
        }

        return in_array($normalized, self::ESTADOS, true) ? $normalized : null;
    }

    public function canBeUpdated(): bool
    {
        return in_array($this->estado, [self::ESTADO_PENDIENTE, self::ESTADO_BORRADOR], true);
    }

    public function canBeEmitted(): bool
    {
        return in_array($this->estado, [self::ESTADO_PENDIENTE, self::ESTADO_BORRADOR], true);
    }

    public function canBeAnnulled(): bool
    {
        return in_array($this->estado, [self::ESTADO_PENDIENTE, self::ESTADO_BORRADOR], true);
    }

    public function cliente()
    {
        return $this->belongsTo(Cliente::class);
    }

    public function user()
    {
        return $this->belongsTo(User::class);
    }

    public function creator()
    {
        return $this->belongsTo(User::class, 'created_by');
    }

    public function updater()
    {
        return $this->belongsTo(User::class, 'updated_by');
    }

    public function emitter()
    {
        return $this->belongsTo(User::class, 'emitida_by');
    }

    public function annuller()
    {
        return $this->belongsTo(User::class, 'anulada_by');
    }

    public function items()
    {
        return $this->hasMany(FacturaItem::class)->orderBy('orden');
    }
}

File: mobile_app/lib/features/facturas/domain/factura.dart
import 'factura_item.dart';

class Factura {
  Factura({
    required this.id,
    required this.codigo,
    required this.numero,
    required this.fecha,
    required this.ciudadExpedicion,
    required this.clienteNombre,
    required this.clienteNit,
    required this.clienteContacto,
    required this.clienteDireccion,
    required this.clienteCiudad,
    required this.observaciones,
    required this.firmaPath,
    required this.firmaNombre,
    required this.firmaCargo,
    required this.firmaEmpresa,
    required this.subtotal,
    required this.ivaTotal,
    required this.total,
    required this.estado,
    required this.items,
    required this.createdBy,
    required this.updatedBy,
    required this.createdByName,
    required this.updatedByName,
    required this.emitidaAt,
    required this.anuladaAt,
    required this.createdAt,
    required this.updatedAt,
  });

  final int id;
  final String codigo;
  final String numero;
  final String fecha;
  final String? ciudadExpedicion;
  final String clienteNombre;
  final String? clienteNit;
  final String? clienteContacto;
  final String? clienteDireccion;
  final String? clienteCiudad;
  final String? observaciones;
  final String? firmaPath;
  final String? firmaNombre;
  final String? firmaCargo;
  final String? firmaEmpresa;
  final String subtotal;
  final String ivaTotal;
  final String total;
  final String estado;
  final List<FacturaItem> items;
  final int? createdBy;
  final int? updatedBy;
  final String? createdByName;
  final String? updatedByName;
  final String? emitidaAt;
  final String? anuladaAt;
  final String? createdAt;
  final String? updatedAt;

  bool get isBorrador {
    final value = estado.trim().toLowerCase();
    return value == 'pendiente' || value == 'borrador';
  }

  bool get isPendiente => isBorrador;
  bool get isEmitida => estado.trim().toLowerCase() == 'emitida';
  bool get isAnulada => estado.trim().toLowerCase() == 'anulada';

  factory Factura.fromJson(Map<String, dynamic> json) {
    return Factura(
      id: _parseInt(json['id']),
      codigo: json['codigo']?.toString() ?? '',
      numero: json['numero']?.toString() ?? json['codigo']?.toString() ?? '',
      fecha: json['fecha']?.toString() ?? '',
      ciudadExpedicion: json['ciudad_expedicion']?.toString(),
      clienteNombre: json['cliente_nombre']?.toString() ?? '',
      clienteNit: json['cliente_nit']?.toString(),
      clienteContacto: json['cliente_contacto']?.toString(),
      clienteDireccion: json['cliente_direccion']?.toString(),
      clienteCiudad: json['cliente_ciudad']?.toString(),
      observaciones: json['observaciones']?.toString(),
      firmaPath: json['firma_path']?.toString(),
      firmaNombre: json['firma_nombre']?.toString(),
      firmaCargo: json['firma_cargo']?.toString(),
      firmaEmpresa: json['firma_empresa']?.toString(),
      subtotal: json['subtotal']?.toString() ?? '0',
      ivaTotal:
          json['iva_total']?.toString() ?? json['impuestos']?.toString() ?? '0',
      total: json['total']?.toString() ?? '0',
      estado: json['estado']?.toString() ?? 'pendiente',
      items: (json['items'] as List<dynamic>? ?? const [])
          .whereType<Map<String, dynamic>>()
          .map(FacturaItem.fromJson)
          .toList(),
      createdBy: _parseNullableInt(json['created_by']),
      updatedBy: _parseNullableInt(json['updated_by']),
      createdByName: json['created_by_name']?.toString(),
      updatedByName: json['updated_by_name']?.toString(),
      emitidaAt: json['emitida_at']?.toString(),
      anuladaAt: json['anulada_at']?.toString(),
      createdAt: json['created_at']?.toString(),
      updatedAt: json['updated_at']?.toString(),
    );
  }
}

int _parseInt(dynamic value, {int fallback = 0}) {
  if (value is int) {
    return value;
  }
  if (value is double) {
    return value.toInt();
  }

  return int.tryParse(value?.toString() ?? '') ?? fallback;
}

int? _parseNullableInt(dynamic value) {
  if (value == null) {
    return null;
  }

  final parsed = _parseInt(value, fallback: -1);
  return parsed == -1 ? null : parsed;
}
